enforcer.go

```diff
@@ -584,10 +584,13 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 	functions := e.fm.GetFunctions()
 	if _, ok := e.model["g"]; ok {
 		for key, ast := range e.model["g"] {
-			if len(e.rmMap) != 0 {
+			// g must be a normal role definition (ast.RM != nil)
+			//   or a conditional role definition (ast.CondRM != nil)
+			// ast.RM and ast.CondRM shouldn't be nil at the same time
+			if ast.RM != nil {
 				functions[key] = util.GenerateGFunction(ast.RM)
 			}
-			if len(e.condRmMap) != 0 {
+			if ast.CondRM != nil {
 				functions[key] = util.GenerateConditionalGFunction(ast.CondRM)
 			}
```

Comment:
```golang
			// g is a conditional role definition
			if len(e.condRmMap) != 0 && ast.CondRM != nil {
				functions[key] = util.GenerateConditionalGFunction(ast.CondRM)
			}
			// or g is a normal role definition
			// g must be a conditional role definition or a normal role definition
			// so equals to `len(e.rmMap) != 0 && ast.RM != nil`
			//    equals to `ast.RM != nil`
			if len(e.rmMap) != 0 && (ast.RM != nil || functions[key] == nil) {
				functions[key] = util.GenerateGFunction(ast.RM)
			}
```

Not sure which one is more readable. `len(e.rmMap)` only indicates that we have at least 1 role definition, so `ast.RM` is better. And `ast.RM != nil` is more clear.

Reply:
@mtt0 Good job! 
`ast.RM != nil` looks more readable, but there are some tests that don't pass at the moment, please fix this.
![image](https://github.com/casbin/casbin/assets/46661603/b6fa77d0-63e4-48ee-ae94-3165ab6d2a50)

It looks like this is when our `AddDef` didn't update `ast.RM` and `ast.CondRM`, you can try to update them in `AddDef` like `initRmMap`

```golang
	if sec == "r" || sec == "p" {
		ast.Tokens = strings.Split(ast.Value, ",")
		for i := range ast.Tokens {
			ast.Tokens[i] = key + "_" + strings.TrimSpace(ast.Tokens[i])
		}
	} else if sec == "g" {
		ast.ParamsTokens = getParamsToken(ast.Value)
		ast.Tokens = strings.Split(ast.Value, ",")
		ast.Tokens = ast.Tokens[:len(ast.Tokens)-len(ast.ParamsTokens)]
		if len(ast.Tokens) <= 2 && len(ast.ParamsTokens) == 0 {
			ast.RM = defaultrolemanager.NewRoleManagerImpl(10)
		} else if len(ast.Tokens) <= 2 && len(ast.ParamsTokens) != 0 {
			ast.CondRM = defaultrolemanager.NewConditionalRoleManager(10)
		} else if len(ast.Tokens) > 2 && len(ast.ParamsTokens) == 0 {
			ast.RM = defaultrolemanager.NewRoleManager(10)
		} else if len(ast.Tokens) > 2 && len(ast.ParamsTokens) != 0 {
			ast.CondRM = defaultrolemanager.NewConditionalDomainManager(10)
		}
	} else {
		ast.Value = util.RemoveComments(util.EscapeAssertion(ast.Value))
	}
```

By the way, if you are convenient, you can add some test cases of different types of role definition, thank you very much for your contribution!

Reply:
Done. I tried insert codes below:

```golang
		if len(ast.Tokens) <= 2 && len(ast.ParamsTokens) == 0 {
			ast.RM = defaultrolemanager.NewRoleManagerImpl(10)
		}
		if len(ast.Tokens) <= 2 && len(ast.ParamsTokens) != 0 {
			ast.CondRM = defaultrolemanager.NewConditionalRoleManager(10)
		}
		if len(ast.Tokens) > 2 {
			if len(ast.ParamsTokens) == 0 {
				ast.RM = defaultrolemanager.NewRoleManager(10)
			} else {
				ast.CondRM = defaultrolemanager.NewConditionalDomainManager(10)
			}
			matchFun := "keyMatch(r_dom, p_dom)"
			if strings.Contains(model["m"]["m"].Value, matchFun) {
				ast.RM.AddDomainMatchingFunc("g", util.KeyMatch)
			}
		}
```

after `AddDef` L90:
https://github.com/casbin/casbin/blob/e46d08ee10207a3fd89972e40590f794612e27d1/model/model.go#L87C1-L91.

It works fine. But I prefer to try to make a function do only one thing, because the code is similar, if you modify one of them in the future, it is easy to forget that the other may also need a similar modification. What's more, users must call `e.Enforce` which will finally call `e.initRmMap`, but users may not call `AddDef`.

Reply:
That's right, your design is even better.